btcec/schnorr/musig2/context.go

```diff
@@ -0,0 +1,280 @@
+	signers []*btcec.PublicKey, shouldSort bool) (*Context, error) {
+
+	// As a sanity check, make sure the signing key is actually amongst the sit
+	// of signers.
+	//
+	// TODO(roasbeef): instead have pass all the _other_ signers?
+	pubKey, err := schnorr.ParsePubKey(
+		schnorr.SerializePubKey(signingKey.PubKey()),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var keyFound bool
+	for _, key := range signers {
+		if key.IsEqual(pubKey) {
+			keyFound = true
+			break
+		}
+	}
+	if !keyFound {
+		return nil, ErrSignerNotInKeySet
+	}
+
+	// Now that we know that we're actually a signer, we'll generate the
+	// key hash finger print and second unique key index so we can speed up
+	// signing later.
+	keysHash := keyHashFingerprint(signers, shouldSort)
+	uniqueKeyIndex := secondUniqueKeyIndex(signers, shouldSort)
+
+	// Next, we'll use this information to compute the aggregated public
+	// key that'll be used for signing in practice.
+	combinedKey := AggregateKeys(
+		signers, shouldSort, WithKeysHash(keysHash),
+		WithUniqueKeyIndex(uniqueKeyIndex),
+	)
+
+	return &Context{
+		signingKey:     signingKey,
+		pubKey:         pubKey,
+		keySet:         signers,
+		combinedKey:    combinedKey,
+		uniqueKeyIndex: uniqueKeyIndex,
+		keysHash:       keysHash,
+		shouldSort:     shouldSort,
+	}, nil
+}
+
+// CombinedKey returns the combined public key that will be used to generate
+// multi-signatures  against.
+func (c *Context) CombinedKey() btcec.PublicKey {
+	return *c.combinedKey
+}
+
+// PubKey returns the public key of the signer of this session.
+func (c *Context) PubKey() btcec.PublicKey {
+	return *c.pubKey
+}
+
+// Session represents a musig2 signing session. A new instance should be
+// created each time a multi-signature is needed. The session struct handles
+// nonces management, incremental partial sig vitrifaction, as well as final
+// signature combination. Errors are returned when unsafe behavior such as
+// nonce re-use is attempted.
+//
+// NOTE: This struct should be used over the raw Sign API whenever possible.
+type Session struct {
+	ctx *Context
+
+	localNonces *Nonces
+
+	pubNonces [][PubNonceSize]byte
+
+	combinedNonce [PubNonceSize]byte
+
+	msg [32]byte
+
+	ourSig *PartialSignature
+	sigs   []*PartialSignature
+
+	finalSig *schnorr.Signature
+}
+
+// NewSession creates a new musig2 signing session.
+func (c *Context) NewSession() (*Session, error) {
+	localNonces, err := GenNonces()
+	if err != nil {
+		return nil, err
+	}
+
+	s := &Session{
+		ctx:         c,
+		localNonces: localNonces,
+		pubNonces:   make([][PubNonceSize]byte, 0, len(c.keySet)),
+		sigs:        make([]*PartialSignature, 0, len(c.keySet)),
+	}
+
+	s.pubNonces = append(s.pubNonces, localNonces.PubNonce)
+
+	return s, nil
+}
+
+// PublicNonce returns the public nonce for a signer. This should be sent to
+// other parties before signing begins, so they can compute the aggregated
+// public nonce.
+func (s *Session) PublicNonce() [PubNonceSize]byte {
```

Comment:
Hmm the idea was that you'd only store the context (since it's just the set of keys), and never store (assuming you mean write to disk?) the session. Are there any relevant fields you think should be exported in the context itself?

Reply:
I mean for keeping things in memory on the signing RPC side. We need to keep the session object in memory to be able to add nonces and partial signatures. But maybe we don't need the context after creating it and getting the combined key from it.